flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCheckpointIDCounter.java

```diff
@@ -118,11 +122,46 @@ public void start() throws Exception {
 
                 if (jobStatus.isGloballyTerminalState()) {
                     LOG.info("Removing {} from ZooKeeper", counterPath);
+                    final CompletableFuture<Void> deletionFuture = new CompletableFuture<>();
                     try {
-                        client.delete().inBackground().forPath(counterPath);
+                        client.delete()
+                                .inBackground(
+                                        (curatorFramework, curatorEvent) -> {
+                                            Preconditions.checkArgument(
+                                                    curatorEvent.getType()
+                                                            == CuratorEventType.DELETE,
+                                                    "An unexpected CuratorEvent was monitored: "
+                                                            + curatorEvent.getType());
+                                            Preconditions.checkArgument(
+                                                    counterPath.equals(curatorEvent.getPath()),
+                                                    "An unexpected path was selected for deletion: "
+                                                            + curatorEvent.getPath());
+
+                                            final KeeperException.Code eventCode =
+                                                    KeeperException.Code.get(
+                                                            curatorEvent.getResultCode());
+                                            if (Sets.immutableEnumSet(
+                                                            KeeperException.Code.OK,
+                                                            KeeperException.Code.NONODE)
+                                                    .contains(eventCode)) {
+                                                deletionFuture.complete(null);
+                                            } else {
+                                                deletionFuture.completeExceptionally(
+                                                        KeeperException.create(
+                                                                KeeperException.Code.get(
+                                                                        curatorEvent
+                                                                                .getResultCode())));
```

Comment:
Can you give an example for how this will look like in practice?

Reply:
There are two test cases added to `ZooKeeperCheckpointIDCounterITCase` covering different error types.

Reply:
What does it look like in a log file. I want to know what and how much information is conveyed to the user, and whether that is sufficient to diagnose the issue.